From d1b75e4479a08c1012892700e633659b7b9c5b6a Mon Sep 17 00:00:00 2001
From: Jordan Isaacs <jordan@snowytrees.dev>
Date: Sat, 4 Jan 2025 19:12:05 -0800
Subject: [PATCH] Update eln-cache in correct spot

---
 lisp/startup.el | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/lisp/startup.el b/lisp/startup.el
index 65cf4627b60..342482dd769 100644
--- a/lisp/startup.el
+++ b/lisp/startup.el
@@ -774,9 +774,6 @@ It is the default value of the variable `top-level'."
       (unwind-protect
 	  (command-line)
 
-        (when (featurep 'native-compile)
-          (startup--update-eln-cache))
-
 	;; Do this again, in case .emacs defined more abbreviations.
 	(if default-directory
 	    (setq default-directory (abbreviate-file-name default-directory)))
@@ -1428,6 +1425,11 @@ please check its value")
 	  (setq xdg-dir (concat "~" init-file-user "/.config/emacs/"))
 	  (startup--xdg-or-homedot xdg-dir init-file-user)))
 
+  ;; Amend `native-comp-eln-load-path', since the command line argument,
+  ;; --init-directory may have altered `user-emacs-directory'
+  (when (featurep 'native-compile)
+    (startup--update-eln-cache))
+
   ;; Run the site-start library if it exists.
   ;; This used to come after the early init file, but was moved here to
   ;; make it possible for sites to do early init things on behalf of
